### DISCUSSION TITLE
怕把作業改爛要怎麼解決？

### DISCUSSION THREAD

**Discussion by enoch-lightning** [1101-GS3073-C/Python-Community-GS3073-C#295] [Q&A] [answered]
第一次的作業是用flask的python框架(還是python的flask框架？)，搭配靜態檔案-CSS, javascript與圖片撰寫網頁。我有一個問題:就是我可以保留最原本老師給的檔案，但同時也把我做的檔案git push上去，讓這兩部份完全分離嗎？這樣我如果覺得我把作業做壞了，也可以再次git clone老師原本給的檔案再做一次。

**Comment by fatsoap** [comment, marked as answer]
可以開新的 git branch
假設原本的分支是 main (原本老師給的檔案)，使用指令 ```git checkout -b dev``` 可以從 main 開一個新的分支 dev 並且切換到dev分支(-b 代表開新的分支)
開發時都使用這個分支做改動&commit，要推到github 的話指令是 ```git push origin dev``` ，
可以用 ```git checkout main``` & ```git checkout dev```  切換兩個分支。
如果做壞了，可以切換到 main ```git checkout main``` 然後再從 main 開新的分支 ```git checkout -b dev2```

**Reply by fatsoap** [reply]
開發完後，在將 dev 合併回 main 當中 (可以發 pull request 或是切到 main 分支然後 ```git merge dev```)

或是開分支給原本老師給的檔案，然後在 main 做開發

**Reply by enoch-lightning** [reply]
原來有這個解決方法。不過我目前還看不懂。

**Reply by enoch-lightning** [reply]
我在想，如果不是像你這樣做，就是我什麼東西都沒設定，還是git push到作業的儲存庫，如果改爛了，是不是再領取一次作業就可以了？

**Reply by chia-kai-chang** [reply]
@fatsoap 很精彩的解說。確實，如果要參與別人的專案(或自己的專案)，可能是為了開發一個「A功能」，可以開一個新branch，並命名為「dev-A功能」或「function-A功能」，等開發好之後，再發起Pull Requst，要求別人把這個「dev-A功能」或「function-A功能」的branch合併回main branch上面。

**Reply by enoch-lightning** [reply]
> 我在想，如果不是像你這樣做，就是我什麼東西都沒設定，還是git push到作業的儲存庫，如果改爛了，是不是再領取一次作業就可以了？

那這樣做會成功嗎？這樣就不用使用git。

**Reply by enoch-lightning** [reply]
@fatsoap 想請問你打的這些指令，如`git checkout -b dev`是要在哪裡打，並執行呢？我不知道，謝謝你。

**Reply by fatsoap** [reply]
> > 我在想，如果不是像你這樣做，就是我什麼東西都沒設定，還是git push到作業的儲存庫，如果改爛了，是不是再領取一次作業就可以了？
> 
> 那這樣做會成功嗎？這樣就不用使用git。

這我不確定ㄝ，沒試過

**Reply by fatsoap** [reply]
> @fatsoap 想請問你打的這些指令，如`git checkout -b dev`是要在哪裡打，並執行呢？我不知道，謝謝你。

在命令提示字元， cd 到資料夾內，然後打這些指令

windows 電腦 也可以打開專案的資料夾，右鍵選擇 Git bash here 然後打指令

**Reply by enoch-lightning** [reply]
專案的資料夾就是從github做git clone下來的資料夾(資料夾的內容是被git clone的github儲存庫的內容)嗎？

**Reply by fatsoap** [reply]
> 專案的資料夾就是從github做git clone下來的資料夾(資料夾的內容是被git clone的github儲存庫的內容)嗎？

對

**Reply by enoch-lightning** [reply]
謝謝。:heart:

**Reply by enoch-lightning** [reply]
因為我是直接將作業git clone下來，並直接改git clone下來的資料夾的內容。這個git clone下來的資料夾是從master做git clone下來的，因此這個git clone下來的資料夾也是master，因此我已經來不及使用`git checkout -b devA`開啟一個新的「分支(branch)」`devA`開發，而是已經直接改變master這個branch，因為我是改由master做git clone下來的資料夾，而github的master與我電腦的master是綁在一起的，動一個就是動另一個。
- master是名字為master的「分支(branch)」
想請問我上面的理解正確嗎？謝謝。:heart:

**Comment by enoch-lightning** [comment]
想請問老師：我們的作業有要求直接git push到master這個「分支(branch)」嗎？還是使用@fatsoap提供的方法也可以？謝謝您。

**Reply by chia-kai-chang** [reply]
都可以，在readme說明清楚作品在哪個分支即可。如果沒有說明，就預設在main分支。這個作業不用一板一眼，只要合理說明，過程方法都不限制。

**Reply by enoch-lightning** [reply]
老師我有一個問題：因為我研究了一下github儲存庫的分支(branch)，發現最原始的狀態沒有main，只有master。那master就是main嗎？還是其實兩個是不一樣的，會在不同的地方，謝謝您。
